s3/costs/s3_costs_route.go

```diff
@@ -0,0 +1,178 @@
+//   Copyright 2017 MSolution.IO
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package costs
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/trackit/jsonlog"
+	"github.com/trackit/trackit2/db"
+	"github.com/trackit/trackit2/es"
+	"github.com/trackit/trackit2/routes"
+	"github.com/trackit/trackit2/users"
+	"gopkg.in/olivere/elastic.v5"
+)
+
+// esQueryParams will store the parsed query params
+type esQueryParams struct {
+	dateBegin   time.Time
+	dateEnd     time.Time
+	accountList []uint
+}
+
+// esFilter represents an elasticsearch filter
+type esFilter struct {
+	Key   string
+	Value string
+}
+
+type esFilters = []esFilter
+
+// queryDataTypeToEsFilters represents the different types of data
+// that can be requested from ES and their associated slice of filters
+var queryDataTypeToEsFilters = map[string]esFilters{
+	"storage": esFilters{
+		esFilter{"usageType", "*TimedStorage*"},
+	},
+	"requests": esFilters{
+		esFilter{"usageType", "*Requests*"},
+	},
+	"bandwidthIn": esFilters{
+		esFilter{"usageType", "*In*"},
+		esFilter{"serviceCode", "AWSDataTransfer"},
+	},
+	"bandwidthOut": esFilters{
+		esFilter{"usageType", "*Out*"},
+		esFilter{"serviceCode", "AWSDataTransfer"},
+	},
+}
+
+func init() {
+	routes.MethodMuxer{
+		http.MethodGet: routes.H(getS3CostData).With(
+			db.RequestTransaction{Db: db.Db},
+			users.RequireAuthenticatedUser{},
+			routes.Documentation{
+				Summary:     "get the s3 costs data",
+				Description: "Responds with cost data based on the queryparams passed to it",
+			},
+			routes.QueryArgs{AwsAccountsQueryArg},
+			routes.QueryArgs{DateBeginQueryArg},
+			routes.QueryArgs{DateEndQueryArg},
+		),
+	}.H().Register("/s3_costs")
+}
+
+var (
+	// AwsAccountsQueryArg allows to get the AWS Account IDs in the URL
+	// Parameters with routes.QueryArgs. These AWS Account IDs will be a
+	// slice of Uint stored in the routes.Arguments map with itself for key.
+	AwsAccountsQueryArg = routes.QueryArg{
+		Name:        "accounts",
+		Type:        routes.QueryArgUintSlice{},
+		Description: "The IDs for many AWS account.",
+	}
```

Comment:
Here you make the `AwsAccountsQueryArg` not optional, but in the function `GetS3UsageAndCostElasticSearchParams` line 63 of file `s3/costs/es_request_constructor.go` you write a condition that looks like handling for optional `AwsAccounts`